Tile composition

LostABike opened the pull request:
<!--
Thank you for your contribution to iTwinUI.

If you are only making changes to the docs site using the "Edit page on GitHub" link,
then you can ignore most of this template.
-->

## Changes

<!--
What kind of code changes does this PR include?
Mention anything that could be helpful for reviewers and include screenshots for visual changes.
-->

Convert most Tile's props into polymographic subcomponents with the following structure:
```
    <Tile>
      <Tile.ThumbnailArea>
        <Tile.ThumbnailPicture/>
        <Tile.Badge/>
        <Tile.TypeIndicator/>
        <Tile.QuickAction/>
      </Tile.ThumbnailArea>
      <Tile.Name/>
      <Tile.ContentArea>
        <Tile.Description />
        <Tile.Metadata/>
        <Tile.MoreOptions/>
      </Tile.ContentArea>
      <Tile.Buttons/>
    </Tile>
```

## Testing

<!--
How did you test your changes?
If your PR has visual changes, then make sure they are demonstrated in html test pages as well as
storybook, then approve visual test images for both (`yarn approve:css` and `yarn approve:react`).

If not applicable, you can write "N/A".
-->

- Updated all Storybook.
- Updated all unit tests.

## Docs

<!--
If your PR includes user-facing changes, then update docs in all places (JSDoc, website, stories, etc).
Make sure to include a changeset (`yarn changeset`).

If not applicable, you can write "N/A".
-->
- Included comments for each subcomponents export. Some have example.
- Created example code, but documentation page .mdx might need to be in a separate PR since behaviors differ from Storybook on certain code.

LostABike commented:
@mayank99 can you briefly check if I'm going the right direction? before I try to complete the other subcomponents. Also I made a different file to work on incrementally rather than modifying existing Tile, easier for me this way. Once everything works right I'll deleted old Tile and rename new Tile

mayank99 commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,177 @@
+  /**
+   * Status of the tile.
+   */
+  status?: 'positive' | 'warning' | 'negative';
```
The main Tile needs to have all the high-level state props, including (in addition to `status`): `isSelected`, `isNew`, `variant`, `isActionable`, `isLoading` and `isDisabled`

LostABike replied:
Added these

mayank99 commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,177 @@
+        thumbnail
+      )}
+      {leftIcon &&
+        React.cloneElement(leftIcon as React.ReactElement, {
```
most of these dom elements should be subcomponents too. I expect the final markup to look something like:
```
<Tile>
  <Tile.ThumbnailArea>
    <Tile.Thumbnail />
    <Tile.QuickAction>⭐️</Tile.QuickAction>
    <Tile.QuickAction>💬</Tile.QuickAction>

    <Tile.Badges>
     <Badge>badge 1</Badge>
     <Badge>badge 2</Badge>
    </Tile.Badges>
  </Tile.ThumbnailArea>

  <Tile.Name>Hello</Tile.Name>

   ...
</Tile>
```

in general, we want to avoid `cloneElement` and prefer making new subcomponents

LostABike replied:
I'm unsure how we can avoid `cloneElement` in cases where the Element supplied by user have many props or components within it. For example a thumbnail (this is made up, not real use of these components):
```
    <Avatar
      size='x-large'
      status='online'
      abbreviation='TR'
      backgroundColor={getUserColor('Terry Rivers')}
      image={
        <img src='https://itwinplatformcdn.azureedge.net/iTwinUI/user-placeholder.png' />
      }
      title='Terry Rivers'
        <IconButton styleType='borderless'>
            <SvgInfo />
        </IconButton>
    />
```
And this could be something the user made just for the thumbnail picture, which I think `cloneElement` is the best use here anyway?

LostABike replied:
So based on the above, my thumbnail subcomponent now looks like this:

```
const TileThumbnail = React.forwardRef((props, ref) => {
  const { as: Element = 'div', className, thumbnail, ...rest } = props;
  return (
    <Element
      className={cx('iui-tile-thumbnail', className)}
      ref={ref}
      {...rest}
    >
      {thumbnail && typeof thumbnail === 'string' ? (
        <div
          className='iui-tile-thumbnail-picture'
          style={{ backgroundImage: `url(${thumbnail})` }}
        />
      ) : thumbnail && (thumbnail as JSX.Element).type === 'img' ? (
        React.cloneElement(thumbnail as JSX.Element, {
          className: 'iui-tile-thumbnail-picture',
        })
      ) : React.isValidElement(thumbnail) ? (
        React.cloneElement(thumbnail, {
          className: cx('iui-thumbnail-icon', thumbnail.props.className),
        })
      ) : (
        thumbnail
      )}
    </Element>
  );
}) as PolymorphicForwardRefComponent<'div', TileThumbnailOwnProps>;
```

Do you think it should be broken down even more? But either ways where it is a custom element, there would be a `cloneElement` regardless, no?

LostABike replied:
So even if we make a new subcomponent for QuickAction, it would also return a `cloneElement` of whatever is passed in, like how we have it originally. Or if you can show me an example of how to break that down without cloning? a use case: 
```
      <Tile.QuickAction
          leftIcon={
            <IconButton styleType='borderless'>
              <SvgInfo />
            </IconButton>
          }
       />
```

LostABike replied:
Turned everything into subcomponents

mayank99 commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,698 @@
+// ----------------------------------------------------------------------------
+// Tile.ThumbnailArea component
+
+const TileThumbnailArea = React.forwardRef((props, ref) => {
+  const { as: Element = 'div', className, children, ...rest } = props;
+  return (
+    <Element
+      className={cx('iui-tile-thumbnail', className)}
+      ref={ref}
+      {...rest}
+    >
+      {children}
+    </Element>
+  );
+}) as PolymorphicForwardRefComponent<'div', TileThumbnailAreaOwnProps>;
+TileThumbnailArea.displayName = 'TileNew.ThumbnailArea';
+
+type TileThumbnailAreaOwnProps = {}; // eslint-disable-line @typescript-eslint/ban-types
```
```suggestion
const TileThumbnailArea = polymorphic.div('iui-tile-thumbnail');
TileThumbnailArea.displayName = 'TileNew.ThumbnailArea';
```

mayank99 commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,698 @@
+   * Description text of the tile.
+   * Gets truncated if it can't fit in the tile.
+   */
+  description?: React.ReactNode;
```
doesn't need to be a separate prop, can just use `children`

mayank99 commented:
> The folder variant is relying on the name prop to make its own iui-tile-name within iui-tile-content, should we still keep the name prop and pass it around the subcomponents? Or restructure css to have to have name outside of content even for folder variant?

We can restructure the html/css. It was kept this way in #1125 for backwards compatibility.

adamhock commented on packages/itwinui-react/src/core/index.ts:
```diff
@@ -295,6 +295,22 @@ export type { TextareaProps } from './Textarea/index.js';
 export type { TileProps } from './Tile/index.js';
 
+export { TileNew } from './Tile/index.js';
+export type { TileNewProps } from './Tile/index.js';
```
All of these types can go under the same set of braces

LostABike replied:
I think I can't because exporting type requires `export type` ?
![image](https://github.com/iTwin/iTwinUI/assets/22462084/ad8f9702-b1be-4ed4-9be6-493ec8595116)

Unless you mean for TileNew stuff to go inside the same braces as original `Tile` right above, then I plan to rename `TileNew` to `Tile` soon and remove one of the set :d

adamhock replied:
![image](https://github.com/iTwin/iTwinUI/assets/78564221/b3172824-8d05-42b3-9321-7b4cec076aec)
I was referring to all of these types to be in the same braces since they're all types and exported from the same file

LostABike replied:
Ah yea that's right. Updated

adamhock replied:
Closing since types do not have to be exported anymore

adamhock commented on packages/itwinui-react/src/core/Tile/index.ts:
```diff
@@ -5,3 +5,4 @@
 export { Tile } from './Tile.js';
 export type { TileProps } from './Tile.js';
 export default './Tile';
+export * from './TileNew';
```
Changing this to `'./TileNew.js'` should fix the `yarn lint` issue in the PR

LostABike replied:
Nice! Thanks for the tip. I updated this.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+import cx from 'classnames';
+import {
+  useSafeContext,
+  // useTheme,
```
This can be removed. Also, you can import `useGlobals` and call that within `TileComponent`

LostABike replied:
Right yea. I forgot to clean up comment. Updated

adamhock replied:
Make sure to call `useGlobals` in the `TileComponent`

LostABike replied:
Oop right. Updated :D

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+import { ProgressRadial } from '../ProgressIndicators/index.js';
+import { DropdownMenu } from '../DropdownMenu/index.js';
+import { IconButton } from '../Buttons/index.js';
+// import { ProgressRadial } from '../ProgressIndicators';
```
These commented imports can be removed

LostABike replied:
Cleaned up!

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+       */
+      variant?: 'default' | 'folder';
+      /**
+       * Whether the tile is expected to be interactable (i.e. `onClick`).
```
interactable wasn't coming up as a word in the spellchecker. I think it can be changed to `interactive`. There's another instance in this file where this change can be made as well

LostABike replied:
Thanks for catching. I updated both spots.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+//</Tile>
+
+// ----------------------------------------------------------------------------
+// Main Tile component
```
This should go after context and before `Tile.Action`

LostABike replied:
Makes sense. I moved Tile up.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+>(undefined);
+TileContext.displayName = 'TileContext';
+
+type TileActionOwnProps = {}; // eslint-disable-line
```
The comment on this line should be changed to the following 

`// eslint-disable-line @typescript-eslint/ban-types`

Also, I think it makes sense for the component props to come before the composition like it does here. Could you change this for the other components which list the props after the component?

LostABike replied:
Yea this makes sense. I moved props to above components. Also updated comment.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+export const TileAction = (
+  props: PolymorphicComponentProps<'a', TileActionOwnProps>,
+) => {
+  const tileContext = useSafeContext(TileContext);
```
This can be changed to the following code 

```tsx
const {setActionable} = useSafeContext(TileContext)
```

That way, `useEffect` only depends on `setActionable` instead of the entire context

LostABike replied:
Thanks. I updated this.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+
+const TileThumbnailAvatar = React.forwardRef((props, ref) => {
+  const { as: Element = 'div', className, children, ...rest } = props;
+  if (React.isValidElement(children)) {
```
I'm not exactly sure what this check is for. Do you mind leaving a comment in the code for what this check is for?

LostABike replied:
Actually want to pull @mayank99 in on this one to ask because I copied the logic from original Tile, which is that the `thumbnail` or `children` in this case is a custom React Element for thumbnail that the user passed in, and we cloned it to render 
![image](https://github.com/iTwin/iTwinUI/assets/22462084/b798cbc2-b9e0-4273-9ab0-930148827afd)

So I think the check is for whether what is passed in `thumbnail` is a custom React Element. As opposed to being just a string to a thumbnail image.

But also, I circumvented `React.cloneElement` by using `React.createElement` in order to keep the same logic but without using `cloneElement`. Should we do it like that? Or if not, what else can we do here?
![image](https://github.com/iTwin/iTwinUI/assets/22462084/24fe51c2-f207-4380-a3ae-d357f95a02b6)

mayank99 replied:
Please avoid using deprecated apis such as `cloneElement` or `createElement`. Writing such code increases complexity and is prone to bugs, while providing little benefit.

Ultimately, what we want to achieve is pretty simple: for both `Tile.ThumbnailPicture` and `Tile.ThumbnailAvatar`, we only need to support _two_ cases. The developer should be able to pass either:
- a string containing the url to the image, or
- a component that renders an `<img>` or `<svg>` or `<div>` with background-image

You can use a [discriminated union](https://www.totaltypescript.com/discriminated-unions-are-a-devs-best-friend) type to narrow the prop types (similar to #1120). Something like:
```
{
  url: string
} |
{
  children: string
}
```

After that, you can render a regular div with your css class (whether it's `iui-tile-thumbnail-picture` or `iui-tile-thumbnail-avatar` or whatever).
```jsx
<div
  className={cx('iui-whatever', className)}
  style={{ backgroundImage: `url()`, ...style }}
  {...rest}
/>
``` 
```jsx
<div
  className={cx('iui-whatever', className)}
  {...rest}
>
  {children}
</div>
```

After that, you can handle all cases (svg vs img vs div) inside the css code.

LostABike replied:
Oop I did not realize `createElement` is deprecated 😬 but thanks for the direction 😅

LostABike replied:
I updated this. Working as it should at least on Storybook 🤔 also I removed the use of ThumbnailAvatar altogether and only use ThumbnailPicture to handle both cases. The user will have to use a `url` prop to pass in the link for their image, or they could pass in their own Avatar/SVG. Did not clean up Avatar code yet, trying to see if this is what we want.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+  return (
+    //todo: styling is all off, need to fix after
+    <Element
+      className={cx('iui-button', 'iui-tile-thumbnail-quick-action', className)}
```
This could be fixed when styling issues are fixed, but I'll comment anyway

If `Element` isn't a `button`, do we still want to add `iui-button` styling? The same goes for `TypeIndicator`

LostABike replied:
This has been updated, it is now whatever <Box/> is and no more `iui-button` classname, there is a new subcomponent for icon which is polymographically default to be `button` with`iui-button-icon`, seems to be working as intended.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+// Tile.Name component
+
+const TileName = React.forwardRef(
+  ({ as: Element = 'div', children, ...rest }, ref) => {
```
I'd prefer if `props` were extracted within the function like all of the other components. Just replace this list with `props` and then extract these in the next line

LostABike replied:
Yea that makes sense. I updated this.

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+const TileMoreOptions = React.forwardRef((props, ref) => {
+  const { as: Element = 'div', className, children, ...rest } = props;
+  const [isMenuVisible, setIsMenuVisible] = React.useState(false);
+  const showMenu = React.useCallback(() => setIsMenuVisible(true), []);
```
nitpicky but can these calls just immediately go inside of `onShow` and `onHide`, leaving no need to make these separate variables?

LostABike replied:
Yup I moved these functions to inline :d

adamhock commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,646 @@
+        onShow={showMenu}
+        onHide={hideMenu}
+        menuItems={(close) =>
+          children.map((option) =>
```
Would it make sense to have `menuItems` be a prop here that's passed into `TileMoreOptions`? That was you can verify that it's of type `JSX.Element[]` and can use `map` properly?

LostABike replied:
Hm yea I was having errors with these. I have instead require the type of `children` to be `JSX.Element[]`. Not sure if this is preferable or not?

adamhock reviewed (commented):
This is great work!

Don't forget to leave a note somewhere to update unit tests and add a changeset.

mayank99 commented:
> moving name out of content for folder variant was getting more complicated than I would like to so I keep the original logic and have `name` under `content` as before, and made specific comment about that for `<Tile.Name>` so user knows where to put it in which case. I could try to do grid if we are keen on moving it out of content, though 😅

@LostABike Can you elaborate on what difficulty/complexity you're facing? It kinda makes sense to use different grids for folder/non-folder tiles.

Keeping `name` and `content` as before would make the whole api inconsistent because some parts are passed through `children` and others through props.

LostABike commented:
> > moving name out of content for folder variant was getting more complicated than I would like to so I keep the original logic and have `name` under `content` as before, and made specific comment about that for `<Tile.Name>` so user knows where to put it in which case. I could try to do grid if we are keen on moving it out of content, though 😅
> 
> @LostABike Can you elaborate on what difficulty/complexity you're facing? It kinda makes sense to use different grids for folder/non-folder tiles. I'd expect you'd be able to use something like this for folder tiles:
> 
> ```css
> grid-template-areas:
>   'thumbnail name'
>   'thumbnail content';
> ```
> 
> Keeping `name` and `content` as before would make the whole api inconsistent because some parts are passed through `children` and others through props. And there is also an accessibility benefit to moving name before thumbnail (see #1125).

I was just trying to achieve this without grid, which is possible just too extra. But yea grid makes sense. I updated the css. Folder looks correct now and Name is moved to outside of Content :D  
![image](https://github.com/iTwin/iTwinUI/assets/22462084/7e581e1d-ae32-47b0-b754-42c5533dc5de)

mayank99 commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  useSafeContext,
+  SvgMore,
+  type PolymorphicComponentProps,
+  type PolymorphicForwardRefComponent,
```
can you move this into a separate `import type` statement?

see https://github.com/iTwin/iTwinUI-react/pull/996

LostABike replied:
I see. Updated

mayank99 commented:
> yea grid makes sense. I updated the css. Folder looks correct now and Name is moved to outside of Content :D
![image](https://user-images.githubusercontent.com/22462084/240634206-7e581e1d-ae32-47b0-b754-42c5533dc5de.png)

Make sure to update html files to reflect the new markup.

Also I think it might need some spacing adjustments. Maybe @FlyersPh9 can help with the exact values. First thing that comes to mind is this:
```css
grid-template:
  'thumbnail name' auto
  'thumbnail content' 1fr / 1fr auto;
row-gap:  calc(var(--iui-size-s) * 0.5);
```

If there's margin set on the tile-name, it might need to be removed.

mayank99 commented:
> Quick Action and Type Indicator icons look different than in original Tile?
![image](https://github.com/iTwin/iTwinUI/assets/22462084/7bb1da9e-1300-4466-9d18-1d90425e6d7c)

This is something fishy going on with the markup. There is a button inside another button.
![](https://github.com/iTwin/iTwinUI/assets/9084735/8ea236a3-0c38-4e16-9987-bceca5a0e382)

Can you fix that?

gretanausedaite commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  const [isMenuVisible, setIsMenuVisible] = React.useState(false);
+
+  return (
+    <DropdownMenu
```
Real use case we are missing: We need to allow users define their own DropdownMenu for moreOptions (At least expose onClick which expands menu).

mayank99 replied:
do you mean the button that opens the menu?

gretanausedaite replied:
Yes, one team wants to load custom options on moreOptions button click so they need onClick event exposed.

LostABike replied:
I added an `onClick` prop, is that what you mean?

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  return (
+    //todo: styling is all off, need to fix after
+    <Element
+      className={cx('iui-button', 'iui-tile-thumbnail-quick-action', className)}
```
Do you mean that we don't use `<Element/>` at all but instead use `<IconButton/>`?

LostABike replied:
I return `Box` for this now, but did you mean to instead return `<IconButton/>` instead?

mayank99 replied:
that works, but now I'm thinking this needs another subcomponent `Tile.IconButton` (name debatable) for the custom styling that's applied on top of `IconButton`.

would be used like this:
```
<Tile.QuickAction>
  <Tile.IconButton>⭐️</Tile.IconButton>
</Tile.QuickAction>
```

LostABike replied:
I added another subcomponent `ButtonIcon` 👀 how's this?

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+        isNew={isNew}
+        status={status}
+      />
+      <span className='iui-tile-name-label'>
```
> every subcomponent should try to render one element

Yes I see that. But it looks like icon is controlled by other props that the user passed into the *main* Tile component, like `isLoading`, `isNew` etc, and is not currently something we offer customizability on but instead provide our own icons for whichever cases, which is why it just renders along with name.

Are you proposing that we create a separate subcomponent for icon and then also offer customizability when none of the icon related prop is passed into main Tile? i.e the props (isNew, isLoading,...) take precedent to render even if the user passed in a customized SVG for icon.

mayank99 replied:
I'm suggesting kinda the opposite. By default the icons will be the ones that make sense for `isLoading`, `isNew`, etc. But if the user passes a custom svg, then it should override the default.

LostABike replied:
Thanks for clarification. I have separated `Name` and `Icon`, should work as you mentioned now.

mayank99 replied:
It's missing the convenience option 
![](https://github.com/iTwin/iTwinUI/assets/9084735/e56a0990-1c0c-4681-8cef-3368a814b4a9)

Also the code from the private `TitleIcon` can be moved inside `TileNameIcon`

LostABike commented on packages/itwinui-react/src/core/index.ts:
```diff
@@ -295,6 +295,24 @@ export type { TextareaProps } from './Textarea/index.js';
 export { Tile } from './Tile/index.js';
```
Yup, updated!

LostABike commented on packages/itwinui-react/src/core/Tile/index.ts:
```diff
@@ -5,3 +5,4 @@
 export { Tile } from './Tile.js';
 export type { TileProps } from './Tile.js';
 export default './Tile';
+export * from './TileNew.js';
```
Thanks. Updated.

LostABike commented on apps/storybook/src/Tile.stories.tsx:
Wait do I still need this? I think I removed it somehow

mayank99 replied:
yes you still need it because `TileProps` can no longer be imported from `@itwin/itwinui-react`

LostABike replied:
Okay I added this back

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+    ...rest
+  } = props;
+
+  useGlobals();
```
Yup. Updated to use `Box` and removed `useGlobals`

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  }
+
+  if (React.isValidElement(children) && children.type == 'SVG') {
+    console.log('true');
```
Odd, I thought I corrected this before push but yea, removed.

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+    //todo: styling is all off, need to fix after
+    <Element
+      className={cx(
+        'iui-button',
```
I return `Box` for this now, but did you mean to instead return `<IconButton/>` instead?

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+    );
+  }
+  if (isSelected) {
+    return <SvgCheckmark className='iui-tile-status-icon' aria-hidden />;
```
Updated

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+
+type TileThumbnailAvatarOwnProps = {}; // eslint-disable-line @typescript-eslint/ban-types
+
+const TileThumbnailAvatar = React.forwardRef((props, ref) => {
```
Yup removed.

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+// Tile.Badge component
+type TileBadgeOwnProps = {}; // eslint-disable-line @typescript-eslint/ban-types
+
+const TileBadge = polymorphic.div('iui-tile-thumbnail-badge-container');
```
Makes sense. I updated the name.

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  Buttons: TileButtons,
+});
+
+export type TileNewProps = PolymorphicComponentProps<'div', TileOwnProps>;
```
Removed/

LostABike commented on packages/itwinui-react/src/core/Tile/TileNew.tsx:
```diff
@@ -0,0 +1,657 @@
+  return null;
+};
+
+export const TileNew = Object.assign(TileComponent, {
```
Oops, updated

LostABike commented:
> > Quick Action and Type Indicator icons look different than in original Tile?
> > ![image](https://user-images.githubusercontent.com/22462084/238721971-7bb1da9e-1300-4466-9d18-1d90425e6d7c.png)
> 
> This is something fishy going on with the markup. There is a button inside another button. ![](https://user-images.githubusercontent.com/9084735/240744723-8ea236a3-0c38-4e16-9987-bceca5a0e382.png)
> 
> Can you fix that?

The styling issue and button you mentioned have been fixed when I used a `div` (`Box`) instead, so I think this is good now? I also left a comment in respond to your feedback on those buttons.

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -2,111 +2,73 @@
  *--------------------------------------------------------------------------------------------*/
-import * as React from 'react';
+
+import React from 'react';
```
can you try to minimize the diff?
![](https://github.com/iTwin/iTwinUI/assets/9084735/5cfb2449-2c90-4a98-bc9b-d04b1dbcc12d)

other than order, the big thing is `import * as React` (which should be preferred over `import React`).

LostABike replied:
Noted, I reorganized the imports

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -148,209 +106,289 @@ type TileProps = {
-        </Box>
-      </Box>
-    );
+//<Tile>
+//  <Tile.Thumbnail>
+//  <Tile.Name>
+//  <Tile.Content>  //has meta data
+//  <Tile.Buttons>
+//</Tile>
```
remove this comment

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -367,25 +405,174 @@ const TitleIcon = ({
-        indeterminate
-      />
+    icon = (
+      <ProgressRadial style={{ height: '100%' }} aria-hidden indeterminate />
```
why inline style `height: 100%`?

LostABike replied:
What should we do about this now?

mayank99 replied:
For now, you'll need to adjust the CSS to fix it. I think changing `svg` to `svg:not(.iui-radial)` might do it?

I'm working on changes to `<ProgressRadial>` which will avoid the need for such hacks.

LostABike commented:
> > yea grid makes sense. I updated the css. Folder looks correct now and Name is moved to outside of Content :D
> > ![image](https://user-images.githubusercontent.com/22462084/240634206-7e581e1d-ae32-47b0-b754-42c5533dc5de.png)
> 
> Make sure to update html files to reflect the new markup.
> 
> Also I think it might need some spacing adjustments. Maybe @FlyersPh9 can help with the exact values. First thing that comes to mind is this:
> 
> ```css
> grid-template:
>   'thumbnail name' auto
>   'thumbnail content' 1fr / 1fr auto;
> row-gap:  calc(var(--iui-size-s) * 0.5);
> ```
> 
> If there's margin set on the tile-name, it might need to be removed.

When you say to update the html file to reflect the new markup, you mean https://itwin.github.io/iTwinUI/css/tile.html right? I think barely any css was changed so everything is still looking like it should. But let me know if you notice any discrepancies.

Also about the folder name spacing, I added a margin-top instead to get it to have the same spacing as before, seems ok?

![image](https://github.com/iTwin/iTwinUI/assets/22462084/3ea50949-d286-40a6-8f34-a8918e281fae)

gretanausedaite reviewed (commented):
Code has some build errors

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -148,244 +106,479 @@ type TileProps = {
+// Tile.Action component
+
+export const TileAction = (
+  props: PolymorphicComponentProps<'a', TileActionOwnProps>,
```
Yup I have updated this

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -148,244 +106,479 @@ type TileProps = {
+          styleType='borderless'
+          size='small'
+          aria-label='More options'
+          onClick={onClick}
```
Updated name

LostABike commented:
Updated PR description. I think I went over all and responded/resolved all of them on my end, so now waiting for reviewers' response next. Build failures fixed. Unit tests updated. HTML markup updated. CSS updated and looks good on my end but might be unconventional/not optimal (folder grid, margin etc). Documentation code added, but documentation itself might need to be in a separate PR.

gretanausedaite commented on .changeset/polite-walls-rush.md:
```diff
@@ -0,0 +1,6 @@
+'@itwin/itwinui-css': major
+---
+
+Updated Tile to be composition of customizable subcomponents
```
Can you separate changeset files for css and react changes? This way you could be more descriptive with what is changed, eg. in css the main change is folder tile changing to display: grid.

LostABike replied:
I created 2 separate changesets

gretanausedaite replied:
changeset description should be useful

gretanausedaite commented on apps/website/src/pages/docs/tile.mdx:
```diff
@@ -14,14 +14,26 @@ import * as AllExamples from '~/examples';
 <p>{frontmatter.description}</p>
 
-<Placeholder componentPageName='core-tile' />
+{/* <Placeholder componentPageName='core-tile' /> */}
```
Remove if not needed anymore or add back if you're not ready to update docs page just yet.

LostABike replied:
Oops, actually not meant to push this file, I have reverted the changes.

gretanausedaite reviewed (commented):
Generally looks good

LostABike commented:
Should we add `aria-label` for ALL returning `<Box/>` of each subcomponent? I only see it for certain specific components in our code base like `<IconButton/>`

mayank99 commented:
> Should we add `aria-label` for ALL returning `<Box/>` of each subcomponent? I only see it for certain specific components in our code base like `<IconButton/>`

No, `aria-label` is generally only needed on icon buttons that are hardcoded by us (e.g. the more options button which otherwise shows up as three dots).

mayank99 commented on apps/storybook/src/Tile.stories.tsx:
```diff
@@ -41,97 +40,109 @@ export default {
-      {...rest}
-    />
+    <Tile {...args}>
+      <Tile.Name>
+        <Tile.NameIcon />
+        <Tile.NameLabel>Stadium</Tile.NameLabel>
+      </Tile.Name>
```
when the label is a simple string and when there is no custom icon, we can just show `<Tile.Name name='Stadium'>`

LostABike replied:
I think for Storybook it is better to keep it this way so that users can toggle the icons, since if only `name` prop is passed in, there will be no icons, and if there is NameIcon subcomponent, it overrides the `name` (I talked about this in another comment below)

mayank99 replied:
can we change the website examples and at least the basic story?

LostABike replied:
yea makes sense. I updated this.

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.QuickAction component
+
+const TileQuickAction = React.forwardRef((props, ref) => {
+  const { className, children, ...rest } = props;
+  return (
+    <Box
+      className={cx('iui-tile-thumbnail-quick-action', className)}
+      ref={ref}
+      {...rest}
+    >
+      {children}
+    </Box>
+  );
+}) as PolymorphicForwardRefComponent<'div'>;
```
```suggestion
const TileQuickAction = polymorphic.div('iui-tile-thumbnail-quick-action');
```

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+  const { className, children, name, ...rest } = props;
+  return (
+    <Box className={cx('iui-tile-name', className)} ref={ref} {...rest}>
+      {name ?? children}
```
yes

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+        isLoading,
+        isActionable,
+        isDisabled,
+        setActionable: () => {},
```
I was getting this error
![image](https://github.com/iTwin/iTwinUI/assets/22462084/172e8c0b-b21e-4cd4-80d0-5ed7ca2c3bfc)

mayank99 replied:
that's because you removed the code that was passed here before.
![](https://github.com/iTwin/iTwinUI/assets/9084735/887d5a2e-6639-43e9-a4b4-c115a476f46a)

LostABike replied:
ohhh. I moved these logic back.

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.Action component
+
+export const TileAction = (props: PolymorphicForwardRefComponent<'a'>) => {
```
o right, I updated this

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -12,100 +13,63 @@ import {
 import { IconButton } from '../Buttons/index.js';
 import { ProgressRadial } from '../ProgressIndicators/index.js';
 
+type TileProps = React.ComponentProps<typeof Tile>;
```
done

LostABike commented on apps/storybook/src/Tile.stories.tsx:
```diff
@@ -41,97 +40,109 @@ export default {
+      </Tile.Name>
+      <Tile.ThumbnailArea>
+        <Tile.BadgeContainer>
+          <Badge backgroundColor='hsl(197, 71%, 83%)'>Badge</Badge>
```
oh cool, updated to 'skyblue' now

LostABike replied:
Done, and also simplified most of them to use `name` prop as well since can't toggle icon on website

LostABike commented on apps/storybook/src/Tile.stories.tsx:
```diff
@@ -22,12 +23,10 @@ import SvgInfo from '@itwin/itwinui-icons-react/cjs/icons/Info';
 import SvgStar from '@itwin/itwinui-icons-react/cjs/icons/Star';
 import SvgTag from '@itwin/itwinui-icons-react/cjs/icons/Tag';
 
-type TileProps = React.ComponentProps<typeof Tile>;
```
Done

LostABike commented on apps/storybook/src/Tile.stories.tsx:
```diff
@@ -22,12 +23,10 @@ import SvgInfo from '@itwin/itwinui-icons-react/cjs/icons/Info';
   component: Tile,
   argTypes: {
-    className: { control: { disable: true } },
+    classNameIcon: { control: { disable: true } },
```
Oops yea, fixed.

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+    status?: 'positive' | 'warning' | 'negative';
+  };
+
+  const TitleIcon = ({
```
I'm getting this build error
```
src/core/Tile/Tile.tsx(281,15): error TS2604: JSX element type 'StatusIcon' does not have any construct or call signatures.
```

So doing this instead:
```
const TileNameIcon = React.forwardRef((props, ref) => {
  const { children, ...rest } = props;
  const { status, isLoading, isSelected, isNew } = useSafeContext(TileContext);

  const StatusIcon = !!status && StatusIconMap[status];
  let icon;

  if (isLoading) {
    icon = <ProgressRadial aria-hidden indeterminate />;
  }
  if (isSelected) {
    icon = <SvgCheckmark aria-hidden />;
  }
  if (isNew) {
    icon = <SvgNew aria-hidden />;
  }
  if (StatusIcon) {
    icon = <StatusIcon aria-hidden />;
  }

  return children || icon ? (
    <Box className='iui-tile-status-icon' ref={ref} {...rest}>
      {children ?? icon}
    </Box>
  ) : null;
}) as PolymorphicForwardRefComponent<'div'>;
TileNameIcon.displayName = 'Tile.NameIcon';
```
also returning empty if nothing is passed in so that there is no empty space/box in front of name when the subcomponent is present (especially on Storybook)

mayank99 replied:
StatusIcon should come before the other conditions

LostABike replied:
Right, I moved it up.

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.MoreOptions component
+type TileMoreOptionsOwnProps = {
+  menuOnClick?: React.MouseEventHandler<HTMLElement>;
```
updated to `onButtonClick`

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,244 +107,499 @@ type TileProps = {
+  return (
+    <Box
+      as='span'
+      className={cx('iui-button-icon', className)}
```
it's different
```js
<Tile.QuickAction>
  <Tile.IconButton>...</Tile.IconButton>
</Tile.QuickAction>
```
vs
```js
<Tile.QuickAction>
  <Tile.ButtonIcon>
    <IconButton>...</IconButton>
  </Tile.ButtonIcon>
</Tile.QuickAction>
```

LostABike replied:
so are we returning the component `IconButton` inside the Tile Subcomponent `<Tile.IconButton>` then?

LostABike replied:
Done

gretanausedaite commented on apps/storybook/src/Tile.stories.tsx:
```diff
@@ -9,11 +9,11 @@ import {
   Tag,
   TagContainer,
   Tile,
+  TileProps,
```
Not exported anymore

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+    isLoading,
+    isActionable: isActionableProp,
+    isDisabled,
+    onClick,
```
cool, so lets remove `onClick` from context, and remove any special handling for `onClick`, then update any leftover actionable examples to explicitly use `Tile.Action` instead of `isActionable` prop.

should we also remove `isActionable` prop?

LostABike replied:
@mayank99 I believe this is what you mean? I moved the action logic out of `Tile.NameLabel` into `Tile.Action`, removing the `onClick` context, so now custom `onClick` if any is passed into `Tile.Action` instead (also pushed):

![image](https://github.com/iTwin/iTwinUI/assets/22462084/18d46ffc-906d-4416-b513-e324f25b8aef)

mayank99 replied:
yes, basically that. You're missing rest props though.

Also, we can set default `as` prop based on `href` vs `onClick`, similar to how it's done in `Alert.Action`

https://github.com/iTwin/iTwinUI/blob/ab07a844f5c2bca1c255adc5502e572c34fc8ede/packages/itwinui-react/src/core/Alert/Alert.tsx#L94-L102

gretanausedaite replied:
> should we also remove `isActionable` prop?

I don't remember if there is a use-case for actionable tile without click action. Maybe users could want that hover effect when clicking on more options menu? We could mark that as deprecated and keep only for classname toggle.

LostABike replied:
marked @deprecated

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
+
```
nitpicky, but this newline can be removed. The fewer line changes the better

LostABike replied:
Done :D

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -21,91 +23,51 @@ import { ProgressRadial } from '../ProgressIndicators/index.js';
 
 const TileContext = React.createContext<
   | {
+      status: 'positive' | 'warning' | 'negative' | undefined;
```
nitpicky, but you can change this to `status?` and removed the `undefined` option

LostABike replied:
Updated!

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.Action component
+
+const TileAction = React.forwardRef((props, forwardedRef) => {
```
nitpicky, but `forwardedRef` can change to `ref` for consistency

mayank99 replied:
forwardedRef is the new preferred name

https://github.com/iTwin/iTwinUI/blob/ab07a844f5c2bca1c255adc5502e572c34fc8ede/scripts/createComponent.js#L71

LostABike replied:
okay, so I updated to `forwardedRef` for all of them

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.ThumbnailArea component
+
+const TileThumbnailArea = polymorphic.div('iui-tile-thumbnail');
```
You can replace `polymorphic.div(` with `polymorphic(` since `div` is the default. This change can be made across the entire file

LostABike replied:
oooh thanks! I updated all of the similar instances.

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+        className={cx('iui-tile-thumbnail-picture', className)}
+        style={{
+          backgroundImage:
+            url && typeof url === 'string' ? `url(${url})` : undefined,
```
`url &&` can be removed since there already is a check for if `url` exists

LostABike replied:
I see. Removed.

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+      )}
+    </Box>
+  );
+}) as PolymorphicForwardRefComponent<'div'>;
```
This component should be forwarded as a `span` instead of a `div`

LostABike replied:
Done

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+        {...rest}
+      >
+        <IconButton
+          as='button'
```
This `as` statement can be removed since the default type for `IconButton` is a `button`

LostABike replied:
Thanks! removed

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+   * `NameIcon` next to name of the tile. Goes under <Tile.Name>
+   * @example
+   * <Tile>
+   * <Tile.Name>
```
nitpicky, but the tabbing is off slightly in this example and in the `NameLabel` example

LostABike replied:
Updated :D

adamhock reviewed (commented):
I have only reviewed the `Tile.tsx` and `Tile.test.tsx` files. I will review the rest of the files and possibly leave more comments shortly

Overall, this is impressive! Great job

mayank99 commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
+// ----------------------------------------------------------------------------
+// Tile.Name component
+type TileNameOwnProps = {
+  name?: string;
```
this is done for convenience so the user doesn't always have to manually pass `TileNameIcon`

LostABike replied:
hmm actually wouldn't having the string as children achieve the same result as having the prop though?

mayank99 replied:
`children` is currently rendered as-is. That means `<Tile.Name>some string</Tile.Name>` doesn't render an icon. I think we want to preserve that behavior, otherwise it would be confusing why `<Tile.Name><>string in a fragment</></Tile.Name>` suddenly starts adding an icon

mayank99 replied:
on that note, you can make this a react node
```suggestion
  name?: React.ReactNode;
```

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,239 +105,463 @@ type TileProps = {
-  return null;
+
+  return children || icon ? (
+    <Box className='iui-tile-status-icon' ref={ref} {...rest}>
```
thanks for catching. I updated this!

FlyersPh9 commented:
On the HTML demo page, the folder that is loading appears to have some issues:
1. The more options button is misaligned to the right.
2. The title & spinner appears to be lower than the rest. 
![Screenshot 2023-06-13 at 4 02 11 PM](https://github.com/iTwin/iTwinUI/assets/849817/9bce7901-476c-4131-b72a-d31a3265f06a)

LostABike commented:
> On the HTML demo page, the folder that is loading appears to have some issues:
> 
> 1. The more options button is misaligned to the right.
> 2. The title & spinner appears to be lower than the rest.
>    ![Screenshot 2023-06-13 at 4 02 11 PM](https://user-images.githubusercontent.com/849817/245590980-9bce7901-476c-4131-b72a-d31a3265f06a.png)

For the options button, it seems to be because I didn't have `iui-tile-description` under `iui-tile-content` like the other demos, so I added that, although it might be more optimal that it should look right regardless whether that class is added, which-

brings us to the main issue, is that I think the way I made my grid portion is....not optimal and very hacky/hard-coded? Currently this setup would make current demos look fine but I'm not sure for universal usages...Maybe you can help?
![image](https://github.com/iTwin/iTwinUI/assets/22462084/09f5c1e4-07a3-44b2-a5ae-537e2aa00da9)

LostABike commented:
> The video tiles with the play icon overlaid above a thumbnail are using the incorrect color.
> 
> **Before:** ![Screenshot 2023-06-13 at 4 05 11 PM](https://user-images.githubusercontent.com/849817/245591545-f4e334d9-d903-468f-ac0e-99173554f594.png)
> 
> **After:** ![Screenshot 2023-06-13 at 4 05 35 PM](https://user-images.githubusercontent.com/849817/245591630-e54aa606-91fd-402d-8d03-c6b621716832.png)

Should be fixed now. Removed an unnecessary css hack.

adamhock commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,240 +104,459 @@ type TileProps = {
+ */
+export const Tile = Object.assign(TileComponent, {
+  /**
+   *ThumbnailArea subcomponent that contains `ThumbnailPicture`, `QuickAction`, `TypeIndicator` or `BadgeContainer`
```
I'd add a space before `ThumbnailArea` on this line

LostABike replied:
done 😄 , and thanks!

adamhock reviewed (approved):
One nitpicky comment. Other than that, make sure the tests pass before you merge!

Fantastic job! 🎉

LostABike commented:
@FlyersPh9 do you want to make a last check on the css/html of this? before I merge (if I manage to pass those image tests that is 😂 )

mayank99 commented on packages/itwinui-css/src/tile/tile.scss:
```diff
@@ -150,9 +154,12 @@ $thumbnail-selector: ':is(.iui-thumbnail-icon, .iui-tile-thumbnail-picture)';
+  grid-template-areas:
+    'thumbnail name'
+    'thumbnail content';
+  grid-template-columns: 3fr 7fr;
+  grid-template-rows: 1.5fr 2.5fr;
```
Since @FlyersPh9 is out, here's my next best suggestion for the grid:
```suggestion
  grid-template-columns: 1fr 2fr;
  grid-template-rows: auto 1fr;
```

the columns preserve the 1:2 ratio from before, while the rows will prefer to keep the "name" part fixed and let the "content" part expand.

you will likely need to add some gap between the name and content. I think there was a `margin-bottom` before which you removed. Reintroducing that might fix the visuals.

LostABike replied:
This is working now, as in no more visual issues. Thanks for the suggestion

LostABike commented on packages/itwinui-react/src/core/Tile/Tile.tsx:
```diff
@@ -147,240 +104,459 @@ type TileProps = {
    */
   isDisabled?: boolean;
-  onClick?: React.MouseEventHandler<HTMLElement>;
+  setActionable?: React.Dispatch<React.SetStateAction<boolean>>;
```
Thanks. I removed this

mayank99 reviewed (approved):
LGTM! Nice work
